app/main/viewmanager.ts

```diff
@@ -0,0 +1,128 @@
+		});
+
+		ipcMain.on('view-select', (e: Event, index: number) => {
+			this.select(index);
+		});
+
+		ipcMain.on('view-destroy', (e: Event, index: number) => {
+			this.destroy(index);
+		});
+
+		ipcMain.on('view-destroy-all', () => {
+			this.destroyAll();
+		});
+
+		ipcMain.on('forward-message-view', (e: Event, name: string, ...params: any[]) => {
+			this.views[this.selectedIndex].webContents.send(name, ...params);
+		});
+
+		ipcMain.on('forward-message-all', (e: Event, name: string, ...params: any[]) => {
+			this.forwardMessageAll(name, ...params);
+		});
+
+		ipcMain.on('call-view-function', (e: Event, name: string, ...params: any[]) => {
+			const fn = this.views[this.selectedIndex][name as keyof View];
+			if (typeof fn === 'function') {
+				// Type checking requires spread elements to match up with a rest parameter.
+				// So, using a workaround here.
+				(fn as any)(...params);
+			}
+		});
+	}
+
+	create(props: ViewProps): void {
+		if (this.views[props.index]) {
+			return;
+		}
+		const view = new View(props);
+		this.views[props.index] = view;
+		view.webContents.loadURL(props.url);
+	}
+
+	select(index: number): void {
+		const mainWindow = BrowserWindow.getAllWindows()[0];
+		const view = this.views[index];
+		if (!view || view.isDestroyed()) {
+			console.log('Attempt to select a view that does not exist.');
+			return;
+		}
+		this.selectedIndex = index;
+		if (!view.webContents.getURL()) {
+			const { url } = view;
+			view.webContents.loadURL(url);
+		}
+		mainWindow.setBrowserView(view);
+		view.webContents.focus();
+		this.fixBounds();
+	}
+
+	fixBounds(): void {
+		const SIDEBAR_WIDTH = 54;
+		const view = this.views[this.selectedIndex];
+		const showSidebar = ConfigUtil.getConfigItem('showSidebar', true);
+		if (!view || view.isDestroyed()) {
+			console.log('Attempt to fix bounds for a view that does not exist.');
+			return;
+		}
+		const mainWindow = BrowserWindow.getAllWindows()[0];
+		const { width, height } = mainWindow.getContentBounds();
+
+		view.setBounds({
+			x: showSidebar ? SIDEBAR_WIDTH : 0,
+			y: 0,
+			width: showSidebar ? width - SIDEBAR_WIDTH : width,
+			height
+		});
+		view.setAutoResize({ width: true, height: true });
+	}
+
+	destroy(index: number): void {
+		const mainWindow = BrowserWindow.getAllWindows()[0];
+		const view = this.views[index];
+		if (!view || view.isDestroyed()) {
+			console.log('Attempt to delete a view that does not exist.');
+			return;
+		}
+		if (mainWindow.getBrowserView() === view) {
+			mainWindow.setBrowserView(null);
+		}
+		view.destroy();
+		delete this.views[index];
+	}
+
+	destroyAll(): void {
+		const mainWindow = BrowserWindow.getAllWindows()[0];
+		mainWindow.setBrowserView(null);
+		for (const id in this.views) {
+			this.destroy(this.views[id].index);
+		}
+	}
+
+	forwardMessageAll(name: string, ...args: any[]): void {
+		for (const id in this.views) {
+			this.views[id].webContents.send(name, ...args);
+		}
+	}
+}
+
+export = new ViewManager();
```

Comment:
Do `export default const ViewManager = new ViewManager();` or something else but not this `export =` syntax. We need to stop using this.

Reply:
> I am not sure if ipc calls use much CPU or memory, but I think it would make sense to have the the basic initialisation actions moved to main process.

Yes, the initialization can be moved to main process, but switching tabs, and everything else, will still require ipc Calls.

Reply:
```export default const ViewManager = new ViewManager();``` This doesn't seem to be working. Can you try this?

Reply:
So I made a mistake here, this could be either `export default new ViewManager()` or
```typescript
const viewManager = new ViewManager();
export default viewManager();

// then in index.ts
import ViewManager from './viewmanager';
```

I prefer the second one just because `export default new ViewManager();` looks awkward to me (but that just me).

Reply:
@vsvipul is this being taken care of?